File: python/schema/multi_lang_analyzer.py
import time

from pymilvus import MilvusClient, DataType, Function, FunctionType
import random


if __name__ == "__main__":
    client = MilvusClient(
        uri="http://localhost:19530",
        token="root:Milvus"
    )

    multi_analyzer_params = {
        # Define language-specific analyzers
        # Each analyzer follows this format: <analyzer_name>: <analyzer_params>
        "analyzers": {
            "english": {"type": "english"},  # English-optimized analyzer
            "chinese": {"type": "chinese"},  # Chinese-optimized analyzer
            "default": {"tokenizer": "icu"}  # Required fallback analyzer
        },
        "by_field": "language",  # Field determining analyzer selection
        "alias": {
            "cn": "chinese",  # Use "cn" as shorthand for Chinese
            "en": "english"  # Use "en" as shorthand for English
        }
    }

    # Initialize a new schema
    schema = client.create_schema()

    # Step 2.1: Add a primary key field for unique document identification
    schema.add_field(
        field_name="id",  # Field name
        datatype=DataType.INT64,  # Integer data type
        is_primary=True,  # Designate as primary key
        auto_id=True  # Auto-generate IDs (recommended)
    )

    # Step 2.2: Add language identifier field
    # This MUST match the "by_field" value in language_analyzer_config
    schema.add_field(
        field_name="language",  # Field name
        datatype=DataType.VARCHAR,  # String data type
        max_length=255  # Maximum length (adjust as needed)
    )

    # Step 2.3: Add text content field with multi-language analysis capability
    schema.add_field(
        field_name="text",  # Field name
        datatype=DataType.VARCHAR,  # String data type
        max_length=8192,  # Maximum length (adjust based on expected text size)
        enable_analyzer=True,  # Enable text analysis
        multi_analyzer_params=multi_analyzer_params  # Connect with our language analyzers
    )

    # Step 2.4: Add sparse vector field to store the BM25 output
    schema.add_field(
        field_name="sparse",  # Field name
        datatype=DataType.SPARSE_FLOAT_VECTOR  # Sparse vector data type
    )

    # Create the BM25 function
    bm25_function = Function(
        name="text_to_vector",  # Descriptive function name
        function_type=FunctionType.BM25,  # Use BM25 algorithm
        input_field_names=["text"],  # Process text from this field
        output_field_names=["sparse"]  # Store vectors in this field
    )

    # Add the function to our schema
    schema.add_function(bm25_function)

    # Configure index parameters
    index_params = client.prepare_index_params()

    # Add index for sparse vector field
    index_params.add_index(
        field_name="sparse",  # Field to index (our vector field)
        index_type="AUTOINDEX",  # Let Milvus choose optimal index type
        metric_type="BM25"  # Must be BM25 for this feature
    )

    # Create collection
    COLLECTION_NAME = "multilingual_documents"

    # Check if collection already exists
    if client.has_collection(COLLECTION_NAME):
        client.drop_collection(COLLECTION_NAME)  # Remove it for this example
        print(f"Dropped existing collection: {COLLECTION_NAME}")

    # Create the collection
    client.create_collection(
        collection_name=COLLECTION_NAME,  # Collection name
        schema=schema,  # Our multilingual schema
        index_params=index_params  # Our search index configuration
    )

    # Prepare multilingual documents
    documents = [
        # English documents
        {
            "text": "Artificial intelligence is transforming technology",
            "language": "english",  # Using full language name
        },
        {
            "text": "Machine learning models require large datasets",
            "language": "en",  # Using our defined alias
        },
        # Chinese documents
        {
            "text": "人工智能正在改变技术领域",
            "language": "chinese",  # Using full language name
        },
        {
            "text": "机器学习模型需要大型数据集",
            "language": "cn",  # Using our defined alias
        },
    ]

    # Insert the documents
    result = client.insert(COLLECTION_NAME, documents)
    client.flush(collection_name=COLLECTION_NAME)
    time.sleep(3)

    # Print results
    inserted = result["insert_count"]
    print(f"Successfully inserted {inserted} documents")
    print("Documents by language: 2 English, 2 Chinese")

    # Expected output:
    # Successfully inserted 4 documents
    # Documents by language: 2 English, 2 Chinese

    search_params = {
        "metric_type": "BM25",  # Must match index configuration
        "analyzer_name": "english",  # Analyzer that matches the query language
        "drop_ratio_search": "0",  # Keep all terms in search (tweak as needed)
    }

    # Execute the search
    english_results = client.search(
        collection_name=COLLECTION_NAME,  # Collection to search
        data=["artificial intelligence"],  # Query text
        anns_field="sparse",  # Field to search against
        search_params=search_params,  # Search configuration
        limit=3,  # Max results to return
        output_fields=["text", "language"],  # Fields to include in the output
        consistency_level="Strong",  # Data‑consistency guarantee
    )

    # Display English search results
    print("\n=== English Search Results ===")
    for i, hit in enumerate(english_results[0]):
        print(f"{i + 1}. [{hit.score:.4f}] {hit.entity.get('text')} "
              f"(Language: {hit.entity.get('language')})")

    # Expected output:
    # === English Search Results ===
    # 1. [2.7881] Artificial intelligence is transforming technology (Language: english)

    search_params["analyzer_name"] = "cn"

    chinese_results = client.search(
        collection_name=COLLECTION_NAME,  # Collection to search
        data=["人工智能"],  # Query text
        anns_field="sparse",  # Field to search against
        search_params=search_params,  # Search configuration
        limit=3,  # Max results to return
        output_fields=["text", "language"],  # Fields to include in the output
        consistency_level="Strong",  # Data‑consistency guarantee
    )

    # Display Chinese search results
    print("\n=== Chinese Search Results ===")
    for i, hit in enumerate(chinese_results[0]):
        print(f"{i + 1}. [{hit.score:.4f}] {hit.entity.get('text')} "
              f"(Language: {hit.entity.get('language')})")

    # Expected output:
    # === Chinese Search Results ===
    # 1. [3.3814] 人工智能正在改变技术领域 (Language: chinese)

File: java/src/main/java/org/example/schema/MultiLangAnalyzer.java
package org.example.schema;

import com.google.gson.JsonObject;
import io.milvus.common.clientenum.FunctionType;
import io.milvus.v2.client.ConnectConfig;
import io.milvus.v2.client.MilvusClientV2;
import io.milvus.v2.common.DataType;
import io.milvus.v2.common.IndexParam;
import io.milvus.v2.service.collection.request.AddFieldReq;
import io.milvus.v2.service.collection.request.CreateCollectionReq;
import io.milvus.v2.service.collection.request.DropCollectionReq;
import io.milvus.v2.service.utility.request.FlushReq;
import io.milvus.v2.service.vector.request.InsertReq;
import io.milvus.v2.service.vector.request.SearchReq;
import io.milvus.v2.service.vector.request.data.EmbeddedText;
import io.milvus.v2.service.vector.response.SearchResp;

import java.util.*;

public class MultiLangAnalyzer {
    private static final MilvusClientV2 client;
    static {
        client = new MilvusClientV2(ConnectConfig.builder()
                .uri("http://localhost:19530")
                .build());
    }

    private static void createCollectionWithAnalyzer(Map<String, Object> analyzerParams) {
        CreateCollectionReq.CollectionSchema collectionSchema = CreateCollectionReq.CollectionSchema.builder()
                .build();
        collectionSchema.addField(AddFieldReq.builder()
                .fieldName("id")
                .dataType(DataType.Int64)
                .isPrimaryKey(true)
                .autoID(true)
                .build());
        collectionSchema.addField(AddFieldReq.builder()
                .fieldName("language")
                .dataType(DataType.VarChar)
                .maxLength(255)
                .build());

        collectionSchema.addField(AddFieldReq.builder()
                .fieldName("text")
                .dataType(DataType.VarChar)
                .maxLength(8192)
                .enableAnalyzer(true)
                .multiAnalyzerParams(analyzerParams)
                .build());
        collectionSchema.addField(AddFieldReq.builder()
                .fieldName("sparse")
                .dataType(DataType.SparseFloatVector)
                .build());

        CreateCollectionReq.Function function = CreateCollectionReq.Function.builder()
                .functionType(FunctionType.BM25)
                .name("text_to_vector")
                .inputFieldNames(Collections.singletonList("text"))
                .outputFieldNames(Collections.singletonList("sparse"))
                .build();
        collectionSchema.addFunction(function);

        List<IndexParam> indexes = new ArrayList<>();
        indexes.add(IndexParam.builder()
                .fieldName("sparse")
                .indexType(IndexParam.IndexType.AUTOINDEX)
                .metricType(IndexParam.MetricType.BM25)
                .build());

        client.dropCollection(DropCollectionReq.builder()
                .collectionName("multilingual_documents")
                .build());

        CreateCollectionReq requestCreate = CreateCollectionReq.builder()
                .collectionName("multilingual_documents")
                .collectionSchema(collectionSchema)
                .indexParams(indexes)
                .build();
        client.createCollection(requestCreate);

        System.out.println("Collection created");
    }

    public static void main(String[] args) {
        // english
        Map<String, Object> analyzerParams = new HashMap<>();
        analyzerParams.put("analyzers", new HashMap<String, Object>() {{
            put("english", new HashMap<String, Object>() {{
                put("type", "english");
            }});
            put("chinese", new HashMap<String, Object>() {{
                put("type", "chinese");
            }});
            put("default", new HashMap<String, Object>() {{
                put("tokenizer", "icu");
            }});
        }});
        analyzerParams.put("by_field", "language");
        analyzerParams.put("alias", new HashMap<String, Object>() {{
            put("cn", "chinese");
            put("en", "english");
        }});
        createCollectionWithAnalyzer(analyzerParams);

        List<String> texts = Arrays.asList(
                "Artificial intelligence is transforming technology",
                "Machine learning models require large datasets",
                "人工智能正在改变技术领域",
                "机器学习模型需要大型数据集"
        );
        List<String> languages = Arrays.asList(
                "english", "en", "chinese", "cn"
        );

        List<JsonObject> rows = new ArrayList<>();
        for (int i = 0; i < texts.size(); i++) {
            JsonObject row = new JsonObject();
            row.addProperty("text", texts.get(i));
            row.addProperty("language", languages.get(i));
            rows.add(row);
        }
        client.insert(InsertReq.builder()
                .collectionName("multilingual_documents")
                .data(rows)
                .build());
        client.flush(FlushReq.builder().collectionNames(Collections.singletonList("multilingual_documents")).build());

        Map<String,Object> searchParams = new HashMap<>();
        searchParams.put("metric_type", "BM25");
        searchParams.put("analyzer_name", "english");
        searchParams.put("drop_ratio_search", 0);
        SearchResp searchResp = client.search(SearchReq.builder()
                .collectionName("multilingual_documents")
                .data(Collections.singletonList(new EmbeddedText("artificial intelligence")))
                .annsField("sparse")
                .topK(3)
                .searchParams(searchParams)
                .outputFields(Arrays.asList("text", "language"))
                .build());

        System.out.println("\n=== English Search Results ===");
        List<List<SearchResp.SearchResult>> searchResults = searchResp.getSearchResults();
        for (List<SearchResp.SearchResult> results : searchResults) {
            for (SearchResp.SearchResult result : results) {
                System.out.printf("Score: %f, %s\n", result.getScore(), result.getEntity().toString());
            }
        }

        searchParams.put("analyzer_name", "cn");
        searchResp = client.search(SearchReq.builder()
                .collectionName("multilingual_documents")
                .data(Collections.singletonList(new EmbeddedText("人工智能")))
                .annsField("sparse")
                .topK(3)
                .searchParams(searchParams)
                .outputFields(Arrays.asList("text", "language"))
                .build());

        System.out.println("\n=== Chinese Search Results ===");
        searchResults = searchResp.getSearchResults();
        for (List<SearchResp.SearchResult> results : searchResults) {
            for (SearchResp.SearchResult result : results) {
                System.out.printf("Score: %f, %s\n", result.getScore(), result.getEntity().toString());
            }
        }

        client.close();
    }
}
